docs/source/scilla-in-depth.rst

```diff
@@ -802,6 +830,233 @@ below, ``Any`` can be of type ``IntX``, ``UintX``, ``String``, ``ByStr20`` or
+``<address contents>`` refers to what the address contains.
+
+The hierarchy of address types is as follows:
+
+- ``ByStr20``: A raw byte string of length 20. The type does not
+  provide any guarantee as to what is located at the
+  address. (Generally, ``ByStr20`` is not regarded as an address type,
+  because it can refer to any byte string of length 20, whether it
+  is meant to represent an address or not.)
+
+- ``ByStr20 with end``: A ``ByStr20`` which, when interpreted as a
+  network address, refers to an address that is in use. An address is
+  in use if it either contains a contract, or if the balance or the
+  nonce of the address is greater than 0. (The balance of an address
+  is the number of Qa held by the address account. The nonce of an
+  address is the number of transactions that have been initiated from
+  that address).
+
+- ``ByStr20 with contract end``: A ``ByStr20`` which, when interpreted
+  as a network address, refers to the address of a contract.
+
+- ``ByStr20 with contract field f1 : t1, field f2 : t2, ... end``: A
+  ``ByStr20`` which, when interpreted as a network address, refers to
+  the address of a contract containing the mutable fields ``f1`` of
+  type ``t1``, ``f2`` of type ``t2``, and so on. The contract in
+  question may define more field than the ones specified in the type,
+  but the fields specified in the type must be defined in the contract.
+
+.. note::
+
+   All addresses in use, and therefore by extension all contract
+   addresses, implicitly define a mutable field ``_balance :
+   Uint128``. For user accounts the ``_balance`` field refers to the
+   account balance.
+
+.. note::
+
+   Address types specifying immutable fields or transitions of a
+   contract are not supported.
+
+Assignability
+-------------
+
+The hierarchy of address types defines a subtype relation, which in
+Scilla is referred to as `assignability`:
+
+- Any type is assignable to itself.
+
+- Any address type ``ByStr20 with ... end`` is assignable to
+  ``ByStr20``, e.g., for comparing equality using ``builtin
+  eq``, or as the ``_recipient`` value of a message.
+
+- Any contract address type ``ByStr20 with contract ... end`` is
+  assignable to ``ByStr20 with end``.
+
+- Any contract address type specifying explict fields ``ByStr20 with
+  contract field f1 : t11, field f2 : t12, ... end`` is assignable to
+  a contract address type specifying a subset of those fields
+  ``ByStr20 with contract field f1 : t21, field f2 : t22, ... end``,
+  provided that ``t11`` is assignable to ``t21``, ``t12`` is
+  assignable to ``t22``, and so on for each field specified in both
+  contract types.
+
+- For ADTs with type parameters such as ``List`` or ``Option``, an ADT
+  ``T t1 t2 ...`` is assignable to ``S s1 s2 ...`` if ``T`` is the
+  same as ``S``, and ``t1`` is assignable to ``s1``, ``t2`` is
+  assignable to ``s2``, and so on.
+
+- A map with key type ``kt1`` and value type ``vt1`` is assignable to
```

Comment:
I'm reluctant to go into the "why" of things, because it opens up discussions about how we should have done things differently. I'd rather stick to the "how" or "what", and be done with it.